File: DFS/22. Generate Parentheses/Solution.java
// https://leetcode.com/problems/generate-parentheses/

/**
By Jiapeng
the code is so elegant!
Runtime: 1 ms, faster than 88.90% of Java online submissions for Generate Parentheses.
Memory Usage: 39.4 MB, less than 87.68% of Java online submissions for Generate Parentheses.
**/

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList();
        dfs(n, n, ans, "");
        return ans;
    }
    
    private void dfs(int left, int right, List<String> ans, String cur){
        if(left > right || left < 0 || right < 0)
            return;
        if(left == 0 && right == 0){
            ans.add(cur);
            return;
        }
        dfs(left-1, right, ans, cur+"(");
        dfs(left, right-1, ans, cur+")");
    }
}

File: DFS/22. Generate Parentheses/Solution.py
# https://leetcode.com/problems/generate-parentheses/

# By Jiapeng
# Runtime: 24 ms, faster than 98.88% of Python3 online submissions for Generate Parentheses.
# Memory Usage: 14.2 MB, less than 29.90% of Python3 online submissions for Generate Parentheses.

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.dfs(n, n, "", ans)
        return ans
    
    def dfs(self, left, right, cur, ans):
        if left > right or left < 0 or right < 0:   # don't forget to set the boundary condition left < 0 and right < 0
            return
        if right == left == 0:
            ans.append(cur)
            return
        left -= 1
        self.dfs(left, right, cur+"(", ans)
        left += 1
        right -= 1
        self.dfs(left, right, cur+")", ans)
        right += 1
